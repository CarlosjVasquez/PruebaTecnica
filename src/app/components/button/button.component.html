<button
  [type]="type"
  class="btn btn-{{ color }} btn-{{ size }}"
  [class.is-btn-disabled]="disabled"
  (click)="onHandleClick()"
  [disabled]="disabled"
>
  {{ label }}
</button>
