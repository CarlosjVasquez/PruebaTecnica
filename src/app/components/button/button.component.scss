@import "../../../styles/variables";

.btn {
    display: inline-block;
    font-weight: 600;
    text-align: center;
    white-space: nowrap;
    vertical-align: middle;
    user-select: none;
    border: 1px solid transparent;
    padding: 0.7rem 1rem;
    font-size: 0.8rem;
    line-height: 1.5;
    border-radius: 0.25rem;
    transition: color 0.1s ease-in-out, background-color 0.1s ease-in-out,
    border-color 0.1s ease-in-out, box-shadow 0.1s ease-in-out;
  
    &-sm {
      padding: map-get($btn-sizes, sm);
      font-size: 0.875rem;
    }
  
    &-md {
      padding: map-get($btn-sizes, md);
      font-size: 1rem;
    }
  
    &-lg {
      padding: map-get($btn-sizes, lg);
      font-size: 1.25rem;
    }
  }
  
  .is-btn-disabled {
    pointer-events: none;
    opacity: 0.5;
    cursor: not-allowed;
  }