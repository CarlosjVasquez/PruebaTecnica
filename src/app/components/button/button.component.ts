import { Component, EventEmitter, Input, Output } from '@angular/core';

@Component({
  selector: 'custom-button',
  standalone: true,
  imports: [],
  templateUrl: './button.component.html',
  styleUrl: './button.component.scss',
})
export class ButtonComponent {
  @Input() color: string = 'primary';
  @Input() size: 'sm' | 'md' | 'lg' = 'md';
  @Input() label: string = 'Button';
  @Input() disabled: boolean | null = false;
  @Input() type: 'button' | 'menu' | 'reset' | 'submit' = 'button';
  @Output() onEventClick: EventEmitter<void> = new EventEmitter();

  onHandleClick() {
    this.onEventClick.emit();
  }
}
