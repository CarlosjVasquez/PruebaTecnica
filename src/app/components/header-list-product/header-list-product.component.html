<div class="header">
  <search (onInputEventemitter)="onHandleSubmit($event)" />
  <custom-button [label]="'Agregar'" (onEventClick)="onAddProduct()" />
</div>
