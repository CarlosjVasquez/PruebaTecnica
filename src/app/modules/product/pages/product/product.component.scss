@import "../../../../../styles/variables";

.title {
  border-bottom: 1px map-get($colors, "secondary") solid;
  text-align: center;
}

.content {
  position: relative;
}

.content-x {
  position: absolute;
  top: 1rem;
  right: 1rem;
}

.close {
  color: map-get($colors, "gray");
  float: right;
  font-size: 28px;
  font-weight: bold;
  cursor: pointer;
}

.close:hover,
.close:focus {
  color: map-get($colors, "dark");
  text-decoration: none;
  cursor: pointer;
}
