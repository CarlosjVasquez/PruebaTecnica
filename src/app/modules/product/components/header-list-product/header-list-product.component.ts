import { Component, EventEmitter, Output } from '@angular/core';
import { ButtonComponent } from '@components/button/button.component';
import { SearchComponent } from '@components/search/search.component';
import { Router } from '@angular/router';

@Component({
  selector: 'header-list-product',
  standalone: true,
  imports: [ButtonComponent, SearchComponent],
  templateUrl: './header-list-product.component.html',
  styleUrl: './header-list-product.component.scss',
})
export class HeaderListProductComponent {
  @Output() onInputEventemitter: EventEmitter<string> = new EventEmitter();
  constructor(private readonly router: Router) {}

  onAddProduct() {
    this.router.navigate(['/products/create']);
  }

  onHandleSubmit(search: string) {
    this.onInputEventemitter.emit(search);
  }
}
