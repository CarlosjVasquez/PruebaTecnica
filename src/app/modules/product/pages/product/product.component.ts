import { Component, OnInit } from '@angular/core';
import { FormProductComponent } from '../../components/form-product/form-product.component';
import { Router } from '@angular/router';
import { Product } from '../../models/product';
import { ProductService } from '../../services/product.service';

@Component({
  selector: 'app-product',
  standalone: true,
  imports: [FormProductComponent],
  templateUrl: './product.component.html',
  styleUrl: './product.component.scss',
})
export class ProductComponent implements OnInit {
  title: string = 'Formulario de Registro';
  edit: boolean = false;

  constructor(private readonly router: Router, private readonly productService: ProductService) {
    if (this.router.url === '/product/update') this.edit = true;
  }

  ngOnInit(): void {
    if (this.edit) {
      this.title = 'Actualizar Producto';
    }
  }

  clearSelectProduct() {
    this.productService.selectProduct(new Product('', '', '', '', '', ''));
    this.router.navigate(['/products']);
  }

  onSubmitForm(product: Product) {
    this.edit
      ? this.productService.updateProduct(product)
      : this.productService.addProduct(product);
    this.clearSelectProduct();
  }
}
