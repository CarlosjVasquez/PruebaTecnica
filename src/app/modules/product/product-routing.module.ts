import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { ListProductComponent } from './pages/list-products/list-product.component';
import { ProductComponent } from './pages/product/product.component';

const routes: Routes = [
  {
    path: '',
    component: ListProductComponent,
  },
  {
    path: 'create',
    component: ProductComponent,
  },
  {
    path: 'update',
    component: ProductComponent,
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class ProductRoutingModule { }
