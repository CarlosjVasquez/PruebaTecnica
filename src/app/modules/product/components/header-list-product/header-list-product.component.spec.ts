import { ComponentFixture, TestBed } from '@angular/core/testing';

import { HeaderListProductComponent } from './header-list-product.component';
import { By } from '@angular/platform-browser';
import { ButtonComponent } from '../../../../components/button/button.component';
import { InputComponent } from '../../../../components/input/input.component';
import { Router, RouterModule } from '@angular/router';
import { ProductComponent } from '../../pages/product/product.component';

describe('HeaderListProductComponent', () => {
  let component: HeaderListProductComponent;
  let fixture: ComponentFixture<HeaderListProductComponent>;
  let compiled: HTMLElement;
  let router: Router;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      providers: [
        HeaderListProductComponent,
        ButtonComponent,
        ProductComponent,
      ],
      imports: [
        RouterModule.forRoot([
          { path: 'product/create', component: ProductComponent },
        ]),
      ],
    }).compileComponents();

    router = TestBed.inject(Router);
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(HeaderListProductComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should button', () => {
    const button = fixture.debugElement.query(By.directive(ButtonComponent));
    expect(button).toBeTruthy();
  });

  test('match to snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('should input', () => {
    const input = fixture.debugElement.query(By.directive(InputComponent));
    expect(input).toBeTruthy();
  });

  test('should input', () => {
    const input = fixture.debugElement.query(By.directive(InputComponent));
    expect(input).toBeTruthy();
  });

  test('should navigate to /product/create on add product', () => {
    const navigateSpy = jest.spyOn(router, 'navigate');
    component.onAddProduct();
    expect(navigateSpy).toHaveBeenCalledWith(['/product/create']);
  });

  test('should emit search input event', () => {
    const searchValue = 'test search';
    let emittedSearch: string | undefined;

    component.onInputEventemitter.subscribe((value) => {
      emittedSearch = value;
    });

    component.onHandleSubmit(searchValue);

    expect(emittedSearch).toEqual(searchValue);
  });

  test('should emit search input event from button click', () => {
    const searchValue = 'test search';
    let emittedSearch: string | undefined;

    component.onInputEventemitter.subscribe((value) => {
      emittedSearch = value;
    });

    const button = fixture.debugElement.query(By.directive(ButtonComponent));
    button.triggerEventHandler('click', null);

    fixture.whenStable().then(() => {
      expect(emittedSearch).toEqual(searchValue);
    });
  });
});
