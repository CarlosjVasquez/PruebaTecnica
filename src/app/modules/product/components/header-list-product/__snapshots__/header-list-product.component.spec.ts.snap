// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderListProductComponent match to snapshot 1`] = `
<div
  id="root2"
>
  <div
    class="header"
  >
    <search>
      <custom-input
        id="search"
        name="search"
        placeholder="Search..."
      >
        <div
          class="form-group"
        >
          
          <input
            class="form-input ng-untouched ng-pristine ng-valid"
            id="search"
            max="undefined"
            min="undefined"
            placeholder="Search..."
            type="text"
          />
          
        </div>
      </custom-input>
    </search>
    <custom-button>
      <button
        class="btn btn-md btn-primary"
        type="button"
      >
         Agregar

      </button>
    </custom-button>
  </div>
</div>
`;
